File: src/main/java/leet_code/till_april/ContainerWithMostWater.java
package leet_code.till_april;

public class ContainerWithMostWater {
    public static void main(String[] args) {
        System.out.println(maxArea(new int[]{1,8,6,2,5,4,8,3,7}));
    }


    public static int maxArea(int[] height) {
        int leftP = 0;
        int rightP = height.length - 1;

        int maxArea = 0;

        while (leftP < rightP){

            maxArea = Math.max(maxArea,(rightP-leftP) * Math.min(height[rightP],height[leftP]));

            if (height[leftP] > height[rightP]){ // if left height is greater than right, then there should be chance of
                rightP--;                        // bigger right height in order to get bigger area
            }else {         // and vice-versa
                leftP++;
            }
        }
        return maxArea;
    }
}

File: src/main/java/leet_code/till_april/ContainerWithMostWater.go
package main

import (
	"fmt"
	"math"
)

func main() {
	height := []int{1, 8, 6, 2, 5, 4, 8, 3, 7}
	fmt.Println(maxArea(height))
}

func maxArea(height []int) int {
	l := 0
	r := len(height) - 1
	maxW := math.MinInt
	for l < r {
		maxW = maxWater(maxW, (r-l)*minWater(height[l], height[r]))
		if height[l] > height[r] {
			r--
		} else {
			l++
		}
	}
	return maxW
}

func minWater(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func maxWater(a, b int) int {
	if a > b {
		return a
	}
	return b
}
